pkg/actions/service.go

```diff
@@ -260,29 +331,50 @@ func (s *Service) saveRunManifestObjectStore(ctx context.Context, manifest RunMa
 	return s.Writer.OutputWrite(ctx, storageNamespace, runManifestPath, manifestReader, manifestSize)
 }
 
-func (s *Service) saveRunManifestDB(ctx context.Context, repositoryID graveler.RepositoryID, manifest RunManifest) error {
-	_, err := s.DB.Transact(ctx, func(tx db.Tx) (interface{}, error) {
-		// insert run information
-		run := manifest.Run
-		_, err := tx.Exec(`INSERT INTO actions_runs(repository_id, run_id, event_type, start_time, end_time, branch_id, source_ref, commit_id, passed)
-			VALUES ($1,$2,$3,$4,$5,$6,$7,$8,$9)`,
-			repositoryID, run.RunID, run.EventType, run.StartTime, run.EndTime, run.BranchID, run.SourceRef, run.CommitID, run.Passed)
+func (s *KVService) saveRunManifestDB(ctx context.Context, repositoryID graveler.RepositoryID, manifest RunManifest) error {
+	// TODO(niro): Should we create a transaction?
+	// insert run information
+	run := manifest.Run
+	err := s.storeRun(ctx, protoFromRunResult(&run), repositoryID.String())
+	if err != nil {
+		return err
+	}
+
+	// insert each task information
+	for i := range manifest.HooksRun {
+		hookRun := manifest.HooksRun[i]
+		taskKey := kv.FormatPath(getTasksPath(repositoryID.String(), run.RunID), hookRun.HookRunID)
+		err = s.Store.SetMsgIf(ctx, taskKey, protoFromTaskResult(&hookRun), nil)
 		if err != nil {
-			return nil, fmt.Errorf("insert run information %s: %w", run.RunID, err)
+			return fmt.Errorf("save task result (runID: %s taskKey %s): %w", run.RunID, taskKey, err)
 		}
+	}
 
-		// insert each task information
-		for _, hookRun := range manifest.HooksRun {
-			_, err = tx.Exec(`INSERT INTO actions_run_hooks(repository_id, run_id, hook_run_id, action_name, hook_id, start_time, end_time, passed)
-				VALUES ($1,$2,$3,$4,$5,$6,$7,$8)`,
-				repositoryID, hookRun.RunID, hookRun.HookRunID, hookRun.ActionName, hookRun.HookID, hookRun.StartTime, hookRun.EndTime, hookRun.Passed)
-			if err != nil {
-				return nil, fmt.Errorf("insert run hook information %s/%s: %w", hookRun.RunID, hookRun.HookRunID, err)
-			}
-		}
-		return nil, nil
-	})
-	return err
+	return nil
+}
+
+func (s *KVService) storeRun(ctx context.Context, run *RunResultData, repoID string) error {
+	runKey := getRunPath(repoID, run.RunId)
+	err := s.Store.SetMsg(ctx, runKey, run)
```

Comment:
I think the order here is wrong. 

If you first set the Run and then the secondary indexes, it's much harder to check if you're in inconsistent state. For example, you created the Run, but failed to create the indexes. Getting all runs will return this specific run. To validate that the state is consistent, you would have to look for all secondary indexes, which some are ok to be missing - like with branch index for tags.

If you set the secondary indexes first, the inconsistent state is to have it pointing to a non-existing Run. I think it's easier to validate because (a) the index is pointing at it (b) in most flows you would need to read it regardless.

Reply:
Agree